### DISCUSSION TITLE
Proposal: relative paths in output

### DISCUSSION THREAD

**Discussion by bukzor** [microsoft/pyright#8891] [Ideas]
I find that I've had to do this more than a few times to make pyright's output readable and/or suitable for a pastebin.

```console
$ pyright . | sed -r 's#'"$PWD"'/#./#g'
./devinfra_metrics/json.py
  ./devinfra_metrics/json.py:32:13 - error: Type of parameter "items" is unknown (reportUnknownParameterType)
  ./devinfra_metrics/json.py:32:20 - error: Variable not allowed in type expression (reportInvalidTypeForm)
```

Would you consider a patch to make this the default behavior? I feel like I've even see a helper method somewhere to call `Path.relpath` iff it shortens the `Path`, somewhere...

**Comment by erictraut** [comment]
There are good reasons why pyright outputs absolute paths in its output. Even if this were a desirable change (which I think it is not), it would be relatively disruptive because of existing scripts that do diffs of pyright's output.

If you prefer a different output format, you have the option of using the `--outputjson` option. You can then write a short script (typically just a couple of lines of code) to transform the output any way you'd like.

**Reply by bukzor** [reply]
I accept your decision, but just my two cents:

I'd normally lean the other direction in my tooling -- the `--json` output gives structured, comparable data, and stdout is for ergonomic-but-unstable formatted output.

In other words, I would never have guessed that you'd take a report of stdout instability as a bug. I still can't quite believe it... I'd expect to hear "use --json-output if you need to diff".

But separately, I'm surprised that diff a reason to keep absolute paths. That's always the first thing I need to do when normalizing output for diffing: make all mentioned paths relative.

For anyone passing by, here's the couple of lines you'll need:

```python
from __future__ import annotations

from pathlib import Path
from typing import Generator
from typing import TypedDict

CSI = "\x1b["


class PyrightJsonRangeEndpoint(TypedDict):
    line: int
    character: int


class PyrightJsonRange(TypedDict):
    start: PyrightJsonRangeEndpoint
    end: PyrightJsonRangeEndpoint


class PyrightJsonGeneralDiagnostic(TypedDict):
    file: str
    severity: str
    message: str
    range: PyrightJsonRange
    rule: str


class PyrightJsonSummary(TypedDict):
    filesAnalyzed: int
    errorCount: int
    warningCount: int
    informationCount: int
    timeInSec: float


class PyrightJson(TypedDict):
    version: str
    time: str
    generalDiagnostics: list[PyrightJsonGeneralDiagnostic]
    summary: PyrightJsonSummary


def pyright(*args: str) -> PyrightJson:
    from json import load
    from subprocess import PIPE
    from subprocess import Popen

    pyright = Popen(("pyright", "--outputjson") + args, stdout=PIPE)
    result = pyright.stdout
    assert result
    return load(result)


def pyright_format(pyright_json: PyrightJson) -> Generator[str, None, None]:
    cwd = Path.cwd()
    prev_file = None
    for general_diagnostic in pyright_json["generalDiagnostics"]:
        file = Path(general_diagnostic["file"])
        if file.is_relative_to(cwd):
            file = f"./{file.relative_to(cwd)}"
        else:
            file = str(file)
        if file != prev_file:
            yield f"{file}\n"
        prev_file = file

        line = general_diagnostic["range"]["start"]["line"] + 1
        column = general_diagnostic["range"]["start"]["character"] + 1

        yield f"  {file}:"
        yield f"{CSI}33m{line}{CSI}m:"
        yield f"{CSI}33m{column}{CSI}m - "
        yield f"{CSI}31m"
        yield general_diagnostic["severity"]
        yield f"{CSI}m: "
        yield general_diagnostic["message"].replace("\n", "\n  ")
        yield f" {CSI}90m("
        yield general_diagnostic["rule"]
        yield f"){CSI}m"
        yield "\n"


def main():
    from sys import argv
    from sys import stdout

    pyright_data = pyright(*argv[1:])
    for hunk in pyright_format(pyright_data):
        stdout.write(hunk)


if __name__ == "__main__":
    exit(main())
```

**Comment by cdeil** [comment]
As mentioned in https://github.com/microsoft/pyright/issues/10312 I think it would be great to have a CLI option `reportUseRelativePath` to shorten the output / make it quicker to grok.

**Comment by bukzor** [comment]
For any other passers-by, I've taken to putting this script into my repositories as `bin/pyright`:

```
#!/bin/bash
set -euo pipefail

pyright "$@" |

  # this just makes things so much easier to read
  sed "s,$PWD,.,g" \
;
```

NOTE: this will do the wrong thing if your pwd has a comma or is short enough to be non-unique.

**Comment by herobank110** [comment]
mypy seems to output relative paths
